torch/distributed/pipelining/schedules.py

```diff
@@ -684,43 +698,47 @@ def _step_microbatches(
 
             # Look at the neighboring ranks for this current timestep and determine whether
             # this current rank needs to do any recv communication
-            prev_rank_action = None
-            if time_step < len(prev_rank_ops):
-                prev_rank_action = prev_rank_ops[time_step]
-            if prev_rank_action is not None:
-                computation_type, mb_index, stage_index = prev_rank_action
-                # Only handle sends for the forward from a previous rank
-                if computation_type == _ComputationType.FORWARD:
-                    # If not the last stage, then receive fwd activations
-                    if stage_index != self._num_stages - 1:
-                        # TODO: We are assuming that stage will always receive from stage-1
-                        # however that is not necessarily true of get_fwd_recv_ops
-                        stage = stage_index_to_stage[stage_index + 1]
-                        ops.extend(stage.get_fwd_recv_ops(mb_index))
-                elif computation_type == _ComputationType.BACKWARD:
-                    # Previous rank doing backward has no influence for the current rank forward recv
-                    pass
-                else:
-                    raise ValueError(f"Unknown computation type {computation_type}")
-
-            next_rank_action = None
-            if time_step < len(next_rank_ops):
-                next_rank_action = next_rank_ops[time_step]
-            if next_rank_action is not None:
-                computation_type, mb_index, stage_index = next_rank_action
-                # Only handle receives for the backwards from a next rank
-                if computation_type == _ComputationType.FORWARD:
-                    # Next rank doing forward has no influence for the current rank backward recv
-                    pass
-                elif computation_type == _ComputationType.BACKWARD:
-                    # If not the first stage, then receive bwd gradients
-                    if stage_index != 0:
-                        # TODO: We are assuming that stage will always receive from stage+1
-                        # however that is not necessarily true of get_bwd_recv_ops
-                        stage = stage_index_to_stage[stage_index - 1]
-                        ops.extend(stage.get_bwd_recv_ops(mb_index))
-                else:
-                    raise ValueError(f"Unknown computation type {computation_type}")
+            for prev_rank in all_prev_ranks:
```

Comment:
iiuc, all_prev_ranks for rank 1 would include 
 - rank 0, which is prev in the forward pass
 - rank 2, which is prev in the backward pass

I'm confused in general by the logic here.  i guess i would expect that we have a special case for when 'bwd_prev_rank' is doing a bwd step, then we need to recv from it.  and another special case for when 'fwd_prev_rank' is doing forward, then we need to recv forward input from it.

but it looks like we mix together both prev_ranks into one for loop and then inside the loop we ignore the case where the prev rank is doing backward. 

so i take this to mean that the variable 'all_prev_ranks' really refers to 'prev_forward_rank'?

Reply:
Yeah 'prev_forward_ranks' is a more accurate name. My goal for the logic was, in this current step:

1) look for any forward activations I need from any ranks in the group (`prev_forward_ranks`)
2) look for any backwards gradients I need from any ranks in the groups.

Maybe a clearer name might be something like `sending_activation_ranks` (for the forward ranks) and `sending_gradient_ranks` (for the backward ranks) instead

Reply:
see my other comment (https://github.com/pytorch/pytorch/pull/128976/files#r1647927426) - i wonder if it would be cleaner/safer to 
- first list out the prev-rank and the next-rank, (assuming no skip conns)
- then find any send/recv between self and prev-rank (for either combination of local stages forward or backward) and batch those comms
- same for next-rank